mesonbuild/jobserver.py

```diff
@@ -0,0 +1,339 @@
+# Copyright 2019 The Meson development team
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""An Executor that can act as a GNU Make jobserver client."""
+
+# Based on ThreadPoolExecutor,
+# Copyright 2009 Brian Quinlan. All Rights Reserved.
+# Licensed to PSF under a Contributor Agreement.
```

Comment:
This makes me really, really nervous, can you actually contribute this under the APL?

Reply:
Yes, the Python license is not copyleft and is compatible with the ASL. However, the original copyright license must be preserved.

See for example https://www.gnu.org/licenses/license-compatibility.html.en: "In a combination of programs under lax licenses, each part carries the license it came with. When the code is merged to the point that the parts can't be distinguished any more, that merged code should carry all the licenses of the merged parts. Since all the licenses are lax anyway, this causes no practical problem except that the list of licenses gets long".

Reply:
There have been a few other cases of people submitting code with portions taken from other places (usually Python standard library). They have all been rejected.

Reply:
Was that on legal grounds, or because  cut-and-paste could be a maintainability issue? This is not clear.

If the latter, this code is a drop-in replacement for the standard library's ThreadPoolExecutor, but of course it does something else and cannot be replaced with it.

If the former, it could actually be that I was being overly cautious; The only pieces from the standard library are the head comment ("workers are created as daemon threads...") and exactly ten lines of code elsewhere (87-88, 115-116, 196-201) that can't really be rewritten differently. I can just rephrase the comment and remove the problematic copyright.

Reply:
I have added code under other permissive licenses (MIT, BSD, etc) that's not the problem. The concern I have is:
1) You've put the APL2 header at the top of the file, but it's not clear to me that is correct. Likely you need to have *both* headers, and comment clearly explaining what came from the python, what version of python it came from, and what changes you've made to the copied source, as is required by the PSF license
2) It's not clear from the comment in this file that we can use the code under the terms of the PSF, "Copyright foo, all rights reserved. licensed to the PSF under Contributor Agreement".